File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        
        int[] map = new int[256];
        int count = 0;
        for(int i=0; i<t.length(); i++){
            char ch = t.charAt(i);
            if(map[ch] == 0) count++;
            map[ch]++;
        }
        
        int startIndex=0;
        int minSubstring = Integer.MAX_VALUE;
        
        int i=0;
        int j=0;
        while(j<s.length()){
            
            char ch = s.charAt(j);
            map[ch]--;
            
            if(map[ch] == 0) count--;
            
            if(count == 0){
                
                while(count == 0){
                    
                    int res = j-i+1;
                    if(minSubstring > res){
                        minSubstring = res;
                        startIndex = i;
                    }
                    
                    char c = s.charAt(i);
                    map[c]++;
                    if(map[c] == 1) count++;
                    i++;
                }
            }
            
            j++;
        }
        
        if(minSubstring == Integer.MAX_VALUE) return "";
        return s.substring(startIndex, startIndex+minSubstring);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        
        if(s.length() < t.length()) return "";
        
        int map[256] = {0};
        
        int noOfUniqueChars=0;
        for(int i=0; i<t.length(); i++){
            if(map[t[i]] == 0) noOfUniqueChars++;
            map[t[i]]++;
        }
        
        int minSubstringLength=INT_MAX;
        int startIndex=0;
        int i=0, j=0;
        
        while(j<s.length()){
            
            map[s[j]]--;
            if(map[s[j]] == 0) noOfUniqueChars--;
            
            if(noOfUniqueChars == 0){
                
                while(noOfUniqueChars == 0){
                    
                    int curSubstringLength = j-i+1;
                    if(minSubstringLength > curSubstringLength){
                        minSubstringLength = curSubstringLength;
                        startIndex = i;
                    }
                    
                    map[s[i]]++;
                    if(map[s[i]] == 1) noOfUniqueChars++;
                    i++;
                }
            }
            
            j++;
        }
        
        if(minSubstringLength == INT_MAX) return "";
        return s.substr(startIndex, minSubstringLength);
    }
};